crates/bdk/src/wallet/mod.rs

```diff
@@ -703,116 +648,159 @@ impl<D> Wallet<D> {
+
+        Ok(spk_iter.map(move |(index, spk)| AddressInfo {
+            index,
+            address: Address::from_script(&spk, self.network).expect("valid address script"),
+            keychain,
+        }))
     }
 
-    /// Return a derived address using the specified `keychain` (external/internal).
-    ///
-    /// If `keychain` is [`KeychainKind::External`], external addresses will be derived (used for
-    /// receiving funds).
+    /// Get the next unused address for the given `keychain`, i.e. the address with the lowest
+    /// derivation index that hasn't been used.
     ///
-    /// If `keychain` is [`KeychainKind::Internal`], internal addresses will be derived (used for
-    /// creating change outputs). If the wallet does not have an internal keychain, it will use the
-    /// external keychain to derive change outputs.
+    /// This will attempt to derive and reveal a new address if no newly revealed addresses
+    /// are available. See also [`reveal_next_address`](Self::reveal_next_address).
     ///
-    /// See [`AddressIndex`] for available address index selection strategies. If none of the keys
-    /// in the descriptor are derivable (i.e. does not end with /*) then the same address will
-    /// always be returned for any [`AddressIndex`].
+    /// # Errors
     ///
-    /// # Panics
-    ///
-    /// This panics when the caller requests for an address of derivation index greater than the
-    /// BIP32 max index.
-    fn _get_address(
+    /// If writing to persistent storage fails.
+    pub fn next_unused_address(
         &mut self,
         keychain: KeychainKind,
-        address_index: AddressIndex,
     ) -> Result<AddressInfo, D::WriteError>
     where
         D: PersistBackend<ChangeSet>,
     {
-        let keychain = self.map_keychain(keychain);
-        let txout_index = &mut self.indexed_graph.index;
-        let (index, spk, changeset) = match address_index {
-            AddressIndex::New => {
-                let ((index, spk), index_changeset) = txout_index.reveal_next_spk(&keychain);
-                (index, spk.into(), Some(index_changeset))
-            }
-            AddressIndex::LastUnused => {
-                let ((index, spk), index_changeset) = txout_index.next_unused_spk(&keychain);
-                (index, spk.into(), Some(index_changeset))
-            }
-            AddressIndex::Peek(mut peek_index) => {
-                let mut spk_iter = txout_index.unbounded_spk_iter(&keychain);
-                if !spk_iter.descriptor().has_wildcard() {
-                    peek_index = 0;
-                }
-                let (index, spk) = spk_iter
-                    .nth(peek_index as usize)
-                    .expect("derivation index is out of bounds");
-                (index, spk, None)
-            }
-        };
+        let ((index, spk), index_changeset) = self.indexed_graph.index.next_unused_spk(&keychain);
```

Comment:
I believe this should also use `map_keychain`. My mistake for missing it.

Reply:
Good catch